ReadMe.txt
Программатор SFP и SFP+ by ZemtsovVA v2.3.3.

ПО созданно с целью чтения, а также полного или частичного изменения прошивки SFP и SFP+ модулей.
Возможности ПО:
	1. Чтение и вывод в графический интерфейс значений байтов регистра 0x50 I2C EEPROM SFP или SFP+ модулей в 
	расшифрованном в соответствии с SFF-8472 виде
	2. Изменение значений байтов регистра 0x50 I2C EEPROM SFP или SFP+ модулей в графическом интерйесе в 
	рашифрованном в	соответсвии с SFF-8472 виде
	3. Сохранение считанной прошивки в бинарный файл
	4. Запись измененной в графическом интерфейсе прошивки в регистр 0x50 I2C EEPROM SFP или SFP+ модулей
	5. Загрузка бинарного файла прошивки с расшировкой значений байтов в соответствии с SFF-8472 с 
	возможностью дальнейшего сохранения в бинарный файл или загрузки в регистр 0x50 I2C EEPROM SFP или SFP+ 
	модулей
	6. Ввод пароля для возможности доступа к памяти SFP в режиме записи. Некоторые SFP модули защищены от 
	записи паролем. Для разблокировки памяти необходимо вводить в определенные ячейки памяти пароль. Данное 
	ПО позволяет ввести пароль разблокировки памяти.
	7. Подбор пароля I2C EEPROM SFP или SFP+ модулей из списка (возможно сформировать любой список и вставить
	его в виде тектса с разделителем в виде символа новой строки \n)
	8. Просмотр содержимого раздела A2h памяти SFP (0x51)
	9. Расшифровка значений текущих аварий и параметров SFP

При помощи данного ПО, руководствуясь документом SFF-8472, возможна организация совместимости сторонних SFP 
модулей с брендированным оборудованием, изменение значений серийных номеров SFP, в случае их дублирования, выбор 
режимов работы модуля, в случае неверных заводских значений, изменение параметров DDM и т.д. Более полный список 
возможностей описан в документе SFF-8472. 
P.S. Следует помнить, что прошивка не изменяет физических характеристик модуля, а лишь дает возможность 
задействовать необходимые имеющиеся физические возможности модуля.

Для работы программатора необходимо собрать программатор по схеме в файле Arduino_SFP_Prog_Scheme.vsdx 
(подойдет плата Arduino nano ATmega 328p, WEMOS LOLIN 32 или WEMOS D1 mini).
Предпочтительно использовать WEMOS D1 mini т.к. для нее имеется проект печатной платы, готовой к производству - 
файл SFP_Prog_Board_WEMOS_D1_Mini.lay6

Для платы Wemos LoLin 32 использовать файл ProgrammatorSFP_2.5.ino.lolin32_0x10000.bin, 
в ПО Flash Download Tools https://www.espressif.com.cn/en/support/download/other-tools выбрать чип ESP32
Для платы Wemos D1 mini использовать файл ProgrammatorSFP_2.5.ino.d1_mini_0x01000.bin, 
в ПО Flash Download Tools https://www.espressif.com.cn/en/support/download/other-tools выбрать чип ESP8266
Для платы Arduino Nano использовать прошивку ProgrammatorSFP_2.5.ino.with_bootloader.eightanaloginputs.hex, 
в ПО XLoader https://github.com/binaryupdates/xLoader выбрать чип Duemilanove/Nano(ATmega328)


release_notes.txt
Изменения v1.1:
1. Добавлена возможность записи разных байт по разным адресам;
2. Добавлена возможность выбора типа ввода данных для записи выбранных байт. Варианты "в строку" и "пошагово";
3. Добавлена возможность записи прошивки из файла;
4. Добавлена функция автопересчета контрольной суммы в байтах 64 и 95 с возможностью выбора автоматической записи
контрольной суммы или с подтверждением от пользователя;
5. Добавлен пункт "Настройки";
6. Добавлена возможность ввода парля, пункт меню "Ввести пароль";
7. Добавлена возможность автоввода пароля из списка;
8. Добавлена возможность ввода пути к файлам для скачивания/заливки по умолчанию. При скачивании достаточно
указать имя файла без расширения, оно будет подставленно автоматически. Для выбора папки в текущей папке 
необходимо в начале пути прописать ./, то есть папка firmwares в текущей папке будет задана так: ./firmwares/;
9. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.0.ino.

Изменения v1.2:
1. Добавлена возможность чтения раздела памяти A2h;
2. Добавлена возможность записи раздела A2h;
3. Устранен в вылет в связи с неправильным вводом данных;
4. Добавлена функция автопресчета контрольной суммы для A2h в 95 байте;
5. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.1.ino.

Изменения v1.3:
1. Реализован механизм подбора пролей по словарю или диапазону;
2. Немного ускорена запись байтов благодаря удалению таймаутов;
3. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.2.ino.

Изменения v1.4:
1. Добавлена возможность побайтовой записи (в настройках) для модулей SFP с медленной памятью;
2. Исправлен подборщик паролей, перебирать стал медленне, но точнее;
3. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.3.ino.

Изменения v2.1:
1. Добавлен графический интерфейс!
2. Удалена возможность записи и чтения регистра 0x51 (A2h)
3. За отсутствием необходимости исключена возможность подбора пароля из диапазона. При необходимости
возможно сформировать диапазон в виде столбца в Excel и вставить в окно списка паролей для перебора
4. Исключен файл настроек за отсутствием необходимости
5. Введено логирование в командной строке действий, выполняемых непосредственно с SFP/SFP+ модулем 
(чтение, запись, ввод/подбор пароля)
6. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.4.ino.

Изменения v2.2:
1. Русифицирован интрфей.
2. Добавлено обнаружение доступных адресов I2C EEPROM
3. Для корректной работы этой версии необходима прошивка для Arduino ProgrammatorSFP_2.5.ino.

Изменения v2.3:
1. Добавлена возможность работы с разделом памяти A2h
2. Разсшифровка значений битов флагов аварий
3. Расшифрвка текущих показаний датчиков SFP модуля
4. Реализован выбор темы оформления. Смена темы производится нажатием клавиши F1

Изменеия v.2.3.1:
1. Все в одном окне (реализованы вкладки)
2. Исправлены мелкие баги

Изменения v.2.3.2:
1. Реализована функция записи только измененных байтов прошивки
2. Визуальные дополнения к интерфейсу ПО
3. Автоматическая подгрузка по умолчанию файла PasswordsDict.lst при подборе пароля

Изменения v.2.3.3:
1. Добалена автоконфигурация под оборудование Cisco
2. Добавлена автоконфигурация под оборудование Juniper
3. Устранены мелкие баги

A2h.py
A2h_0_1_desc =  'Высокая температура (Alarm) C'
A2h_2_3_desc =  'Низкая температура (Alarm) C'
A2h_4_5_desc =  'Высокая температура (Warning) C'
A2h_6_7_desc =  'Низкая температура (Warning) C'
A2h_8_9_desc =  'Высокое напряжение (Alarm) В'
A2h_10_11_desc =  'Низкое напряжение (Alarm) В'
A2h_12_13_desc =  'Высокое напряжение (Warning) В'
A2h_14_15_desc =  'Низкое напряженеи (Warning) В'
A2h_16_17_desc =  'Высокий ток (Alarm) мА'
A2h_18_19_desc =  'Низкий ток (Alarm) мА'
A2h_20_21_desc =  'Высокий ток (Warning) мА'
A2h_22_23_desc =  'Низкий ток (Warning) мА'
A2h_24_25_desc =  'Высокая вх.мощность (Alarm) мВт'
A2h_26_27_desc =  'Низкая вх.мощность (Alarm) мВт'
A2h_28_29_desc =  'Высокая вх.мощность (Warning) мВт'
A2h_30_31_desc =  'Низкая вх.мощность (Warning) мВт'
A2h_32_33_desc =  'Высокая исх.мощность (Alarm) мВт'
A2h_34_35_desc =  'Низкая исх.мощность (Alarm) мВт'
A2h_36_37_desc =  'Высокая исх.мощность (Warning) мВт'
A2h_38_39_desc =  'Низкая исх.мощность (Warning) мВт'
A2h_40_55_desc =  'Не определено'
A2h_56_59_desc =  'RX_PWR(4) мкВт'
A2h_60_63_desc =  'RX_PWR(3) мкВт'
A2h_64_67_desc =  'RX_PWR(2) мкВт'
A2h_68_71_desc =  'RX_PWR(1) мкВт'
A2h_72_75_desc =  'RX_PWR(0) мкВт'
A2h_76_77_desc =  'Ток смещения лазера (Slope) мкА'
A2h_78_79_desc =  'Ток смещения лазера (Offset) мкА'
A2h_80_81_desc =  'Мощность TX (Slope) мкВт'
A2h_82_83_desc =  'Мощность TX (Offset) мкВт'
A2h_84_85_desc =  'Температура (Slope) С'
A2h_86_87_desc =  'Температура (Offset) С'
A2h_88_89_desc =  'Напряжение (Slope) мкВ'
A2h_90_91_desc =  'Напряжение (Offset) мкВ'
A2h_92_94_desc =  'Не определено'
A2h_95_desc =  'Контрольная сумма (0-94)'
A2h_96_97_desc =  'Температруа С'
A2h_98_99_desc =  'Напряжение В'
A2h_100_101_desc =  'Ток мА'
A2h_102_103_desc =  'Мощность TX мВт'
A2h_104_105_desc =  'Мощность RX мВт'
A2h_106_109_desc =  'Не определено'

A2h_110_desc = 'Статус/Контроль'
A2h_110 = {
    7: 'Статус лазера',
    6: 'Выключить лазер',
    5: 'Статус контакта RS1',
    4: 'Статус контакта RS0',
    3: 'Rate Select',
    2: 'TX Fault',
    1: 'LOS',
    0: 'Готов к приему',
}

A2h_111_desc =  'Зарезервировано'

A2h_112 = {
    7: 'Высокая теспература Alarm',
    6: 'Низкая температура Alarm',
    5: 'Высокое напряжение Alarm',
    4: 'Низкое напряжение Alarm',
    3: 'Высокий ток Alarm',
    2: 'Низкий ток Alarm',
    1: 'Высокая исх.мощность Alarm',
    0: 'Низкая исх.мощность Alarm',
}

A2h_113 = {
    7: 'Высокая вх.мощность Alarm',
    6: 'Низкая вх мощность Alarm',
    5: 'Reserverd Alarm',
    4: 'Reserverd Alarm',
    3: 'Reserverd Alarm',
    2: 'Reserverd Alarm',
    1: 'Reserverd Alarm',
    0: 'Reserverd Alarm',
}

A2h_114_desc = 'Не определено'
A2h_115_desc = 'Не определено'

A2h_116 = {
    7: 'Высокая теспература Warning',
    6: 'Низкая температура Warning',
    5: 'Высокое напряжение Warning',
    4: 'Низкое напряжение Warning',
    3: 'Высокий ток Warning',
    2: 'Низкий ток Warning',
    1: 'Высокая исх.мощность Warning',
    0: 'Низкая исх.мощность Warning',
}

A2h_117 = {
    7: 'Высокая вх.мощность Warning',
    6: 'Низкая вх мощность Warning',
    5: 'Reserverd Warning',
    4: 'Reserverd Warning',
    3: 'Reserverd Warning',
    2: 'Reserverd Warning',
    1: 'Reserverd Warning',
    0: 'Reserverd Warning',
}

A2h_118 = {
    7: 'Reserved',
    6: 'Reserved',
    5: 'Reserved',
    4: 'Reserved',
    3: 'Программный RATE_SELECT',
    2: 'Reserved',
    1: 'Максимальный уровень мощности Статус',
    0: 'Максимальный уровень мощности',
}

A2h_119_desc = 'Не определено'

Ah2_120_127_desc = 'Пароль'

A2h_128_247_desc = 'Пользовательските данные'

A2h_248_255_desc = 'Данные вендора'

A2h_writeAllow = [
        0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,
        31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,
        59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,
        87,88,89,90,91,92,93,94,110,111,118,119,120,121,122,123,124,125,126,127,128,129,
        130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,
        151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,
        172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,
        193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,
        214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,
        235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255
]






src/bitsOper.py
def bitsDetect(byte_val): #определить, какие биты в байте равны 1
    result = list()
    a = 0x80
    i = 7
    while a >= 1:
        if byte_val/a >= 1: 
            result.append(i)
            byte_val -= a
        a = int(a/2)
        i -= 1
    return result
    
def bitsConvert (lst): #получить байт, в котором биты списка lst равны 1
    result = 0
    a = 0x80
    i = 7
    while i >= 0:
        if i in lst:
            result += a
        a = int(a/2)
        i -= 1
    return result
    
    
def twos_comp(val, bits):
    try:
        val = int(val.hex(),16)
    except:
        pass
    if (val & (1 << (bits - 1))) != 0:
        val = val - (1 << bits)
    val = round((val/256), 3)
    return val
    
    
def de_twos_comp(val,bits):
    if val < 0:
        d = abs(int(val//1))
        c = round(abs(val)%1,3)
        if c>0: d -= 1
        b1 = 255-d
        b2 = round(255-c/(1/256))
        val = (b1<<8) + b2 + 1
        val = int.to_bytes(val,int(bits/8),'big')
    else:
        val = round(val/(1/256))
        val = int.to_bytes(val,int(bits/8),'big')
    return val
    
    
def from_slope(val):
    val = val[0] + val[1]/256
    return round(val,4)
    
def to_slope(val):
    d = int(val//1)
    c = val%1
    val = bytes([d,int(c*256)])
    return val


src/A0h.py
A0h_0_desc = 'Идентификатор интерфейса'
A0h_0 = {
    0x00: 'Неизвестный',
    0x01: 'GIBIC',
    0x02: 'Впаян в материнскую плату',
    0x03: 'SFP или SFP+',
    0x04: '300 pin XBI',
    0x05: 'Xenpak',
    0x06: 'XFP',
    0x07: 'XFF',
    0x08: 'XFP-E',
    0x09: 'XPak',
    0x0a: 'X2',
    0x0b: 'DWDM-SFP',
    0x0c: 'QSFP',
}
for i in range(0x0d,0x7f+1):
    A0h_0[i] = hex(i)+' Не определен'
for i in range(0x80,0xff+1):
    A0h_0[i] = hex(i)+' Определен вендором'

A0h_1_desc = 'Расширенный идентификатор'
A0h_1 = {
    0x00: 'Не ссответсвует MOD_DEF',
    0x01: 'GIBIC MOD_DEF 1',
    0x02: 'GIBIC MOD_DEF 2',
    0x03: 'GIBIC MOD_DEF 3',
    0x04: 'GIBIC/SFP',
    0x05: 'GIBIC MOD_DEF 5',
    0x06: 'GIBIC MOD_DEF 6',
    0x07: 'GIBIC MOD_DEF 7',
}
for i in range(0x08,0xff+1):
    A0h_1[i] = hex(i)+' Не определен'

A0h_2_desc = 'Тип Коннектора'
A0h_2 = {
    0x00: 'Неизвестный',
    0x01: 'SC',
    0x02: 'Медный разъем типа 1',
    0x03: 'Медный разъем типа 2',
    0x04: 'BNC/TNC',
    0x05: 'Коаксиальный (оптика)',
    0x06: 'FiberJack',
    0x07: 'LC',
    0x08: 'MT-RJ',
    0x09: 'MU',
    0x0a: 'SG',
    0x0b: 'Оптический Пигтейл',
    0x0c: 'MPO Parallel Optic',
    0x20: 'HSSDC II',
    0x21: 'Медный Пигтейл',
    0x22: 'RJ45',
}
for i in range(0x23,0x7f+1):
    A0h_2[i] = hex(i)+' Не определен'
for i in range(0x80,0xff+1):
    A0h_2[i] = hex(i)+' Определен вендором'

A0h_3_desc = ['Соответствие 10G Ethernet', 'Соответствие InfiniBand']
A0h_3 = [
    {
    7: '10G Base-ER',
    6: '10G Base-LRM',
    5: '10G Base-LR',
    4: '10G Base-SR',
    },
    {
    3: '1X SX',
    2: '1X LX',
    1: '1X Copper Active',
    0: '1X Copper Passive',
    }
]

A0h_4_desc = 'ESCON/SONET'
A0h_4 = {
    7: 'ESCON MMF, 1310 LED',
    6: 'ESCON SMF, 1310 Laser',
    5: 'OC-192, short',
    4: 'SONET reach specifier bit 1',
    3: 'SONET reach specifier bit 2',
    2: 'OC-48, long',
    1: 'OC-48, inter.',
    0: 'OC-48, short',
}

A0h_5_desc = 'SONET'
A0h_5 = {
    7: 'Unallocated',
    6: 'OC-12, single, long',
    5: 'OC-12, single, inter.',
    4: 'OC-12, short',
    3: 'Unallocated',
    2: 'OC-3, single,long',
    1: 'OC-3, single, inter.',
    0: 'OC-3, short',
}

A0h_6_desc = 'Соответствие Ethernet'
A0h_6 = {
    7: 'BASE-PX',
    6: 'BASE-BX10',
    5: '100BASE-FX',
    4: '100BASE-LX/LX10',
    3: '1000BASE-T',
    2: '1000BASE-CX',
    1: '1000BASE-LX',
    0: '1000BASE-SX',
}

A0h_7_desc = ['Длина соединения (оптика)', 'Оптическая технология']
A0h_7 = [
    {
    7: 'V (very long)',
    6: 'S (short)',
    5: 'I (inter.)',
    4: 'L (long)',
    3: 'M (medium)',
    },
    {
    2: 'Shortwave laser, linear Rx (SA)',
    1: 'Longwave laser (LC)',
    0: 'Electrical inter (EL)',
    }
]

A0h_8_desc = ['Оптическая технология',  'Технология SFP+ кабеля']
A0h_8 = [
    {
    7: 'Electrical intra (EL)',
    6: 'Shortwave laser w/o OFC (SN)',
    5: 'Shortwave laser with OFC (SL)',
    4: 'Longwavw laser (LL)',
    },
    {
    3: 'Активный кабель',
    2: 'Пассивный кабель',
    1: 'Не определено',
    0: 'Не определено',
    }
]

A0h_9_desc = 'Среда передачи'
A0h_9 = {
    7: 'Сдвоенная пара (TW)',
    6: 'Витая пара (TP)',
    5: 'Коаксиальный мини(MI)',
    4: 'Коаксиальный видео (TV)',
    3: 'Мультимод. ОВ 62.5нм (M6)',
    2: 'Мультимод. ОВ, 50нм (M5, M5E)',
    1: 'Не определен',
    0: 'Одномод. ОВ (SM)',
}

A0h_10_desc = 'Скорость передачи'
A0h_10 = {
    7: '1200 МБ/сек',
    6: '800 МБ/сек',
    5: '1600 МБ/сек',
    4: '400 МБ/сек',
    3: 'Не определено',
    2: '200 МБ/сек',
    1: 'Не определено',
    0: '100 МБ/сек',
}

A0h_11_desc = 'Кодировка'
A0h_11 = {
    0x00: 'Не определено',
    0x01: '8B/10B',
    0x02: '4B/5B',
    0x03: 'NRZ',
    0x04: 'Manchester',
    0x05: 'SONET Scrambled',
    0x06: '64B/66B',
}
for i in range(0x07,0xff+1):
    A0h_11[i] = hex(i)+' Не определено'

A0h_12_desc = 'Номинальная скорость'

A0h_13_desc = 'Идентификатор скорости'
A0h_13 = {
    0x00: 'Не определен',
    0x01: 'SFF-8079 (4/2/1G)',
    0x02: 'SFF-8431 (8/4/2G Rx)',
    0x03: 'Не определен',
    0x04: 'SFF-8431 (8/4/2G Tx)',
    0x05: 'Не определен',
    0x06: 'SFF-8431 (8/4/2G Rx&Tx)',
    0x07: 'Не определен',
    0x08: 'FC-PI-5 (10/8/4G Rx,Tx)',
    0x0a: 'Не определен',
}
for i in range(0x0b,0xff+1):
    A0h_13[i] = hex(i)+' Не определен'

A0h_14_desc = 'Длина (SM) км'

A0h_15_desc = 'Длина (SM) 100м'

A0h_16_desc = 'Длина (50um, OM2) 10м'

A0h_17_desc = 'Длина (65.2um, OM1) 10м'

A0h_18_desc = 'Длина (Медь) м'

A0h_19_desc = 'Длина (50um, OM3) 10м'

A0h_vendorName = [20 - 35]

A0h_vendorOUI = [37-39]

A0h_vendorPN = [40-55]

A0h_vendorRev = [56-59]

#------------------------------------------------
#---Optical or Passive/active cable (A0h_3 to 10)
#------------------------------------------------

A0h_60_desc = 'Длина волны 1'

A0h_61_desc = 'Длина волны 2'

A0h_60_passive_desc = 'Passive Cabble Spec. 1'

A0h_61_passive_desc = 'Passive Cable Spec. 2'

A0h_60_passive = {
    7: 'Unallocated',
    6: 'Unallocated',
    5: 'SFF-8461',
    4: 'SFF-8461',
    3: 'SFF-8461',
    2: 'SFF-8461',
    1: 'FC-PI-4 Appendix H',
    0: 'SFF-8431 Appendix E',
}

A0h_61_passive = {
    7: 'Unallocated',
    6: 'Unallocated',
    5: 'Unallocated',
    4: 'Unallocated',
    3: 'Unallocated',
    2: 'Unallocated',
    1: 'Unallocated',
    0: 'Unallocated',
}

A0h_60_active_desc = 'Active Cabble Spec. 1'

A0h_61_active_desc = 'Active Cable Spec. 2'

A0h_60_active = {
    7: 'Unallocated',
    6: 'Unallocated',
    5: 'Unallocated',
    4: 'Unallocated',
    3: 'FC-PI-4',
    2: 'SFF-8431',
    1: 'FC-PI-4 Appendix H',
    0: 'SFF-8431 Appendix E',
}

A0h_61_active = {
    7: 'Unallocated',
    6: 'Unallocated',
    5: 'Unallocated',
    4: 'Unallocated',
    3: 'Unallocated',
    2: 'Unallocated',
    1: 'Unallocated',
    0: 'Unallocated',
}

#----------------------------------------------------
#---End Optical or Passive/active cable (A0h_3 to 10)
#----------------------------------------------------

A0h_63_desc = 'Контрольная сумма 0-62'

A0h_64_desc = 'Опции 1'

A0h_64 = {
    7: 'Не определено',
    6: 'Не определено',
    5: 'Не определено',
    4: 'Не определено',
    3: 'Не определено',
    2: 'Охлаждаемый лазер',
    1: 'Ур. мощн. (0 - Уровень 1, 1 - 2)',
    0: 'Линеный выход премника',
}

A0h_65_desc = 'Опции 2'

A0h_65 = {
    7: 'Не определено',
    6: 'Не определено',
    5: 'Поддержка RATE_SELECT',
    4: 'Поддержка TX_DISABLE',
    3: 'Поддержка TX_FAULT',
    2: 'Поддержка Обнаружения синала',
    1: 'Поддержка LOS',
    0: 'Не определено',
}

A0h_66_desc = 'Скорость max'

A0h_67_desc = 'Скорость min'

A0h_vendorSN = [68-83]

A0h_dataCode = [84-91]

A0h_92_desc = 'Тип диагностики и мониторинга'

A0h_92 = {
    7: 'Устаревший (Legacy)',
    6: 'DDM',
    5: 'Внутренняя Калибровка',
    4: 'Внешняя Калибровка',
    3: 'Вх. мощн.(0 - OMA, 1 - Ср. знач.)',
    2: 'Требуется изменение адреса',
    1: 'Не определен',
    0: 'Не определен',
}

A0h_93_desc = 'Расширенные опции'

A0h_93 = {
    7: 'Поддержка флага Alarm/Warning',
    6: 'Программное управление TX_DISABLE',
    5: 'Программный мониторинг TX_FAULT',
    4: 'Программный мониторинг RX_LOS',
    3: 'Программное управление RATE_SELECT',
    2: 'Application Select control (SFF-8079)',
    1: 'Soft Rate Select control (SFF-8431)',
    0: 'Не определено',
}

A0h_94_desc = 'Соответствие SFF-8472'

A0h_94 = {
    0x00: 'Неизвестно',
    0x01: 'Rev.9.3 of SFF-8472',
    0x02: 'Rev.9.5 of SFF-8472',
    0x03: 'Rev.10.2 of SFF-8472',
    0x04: 'Rev.10.4 of SFF-8472',
    0x05: 'Rev.11.0 of SFF-8472',
}
for i in range(0x06,0xff+1):
    A0h_94[i] = hex(i)+' Не определено'
    
A0h_95_desc = 'Контрольная сумма 64-94'

